uro98/assignment2/PrintAncestors.java

```diff
@@ -0,0 +1,84 @@
+import java.util.Arrays;
+import java.util.LinkedList;
+
+public class PrintAncestors {
+  
+  /**
+   * Print the ancestors of the node with the given key in a binary tree as a list.
+   * Uses left-to-right depth first search.
+   * An empty list is printed if the node is not found.
+   * @param tree  a generic binary tree
+   * @param key   the target value of a node
+   */
+  public static <T> void printAncestors(BinaryTree<T> tree, T key) {
+    LinkedList<T> ancestors = new LinkedList<T>();
+    LinkedList<Node<T>> frontier = new LinkedList<Node<T>>();
```

Comment:
Optional: type of variable could be `Deque<Node<T>>` to make it more explicit that it is used as a queue.

Reply:
I'm using `frontier` as a stack though, should I still change it to `Deque<Node<T>>`?

Reply:
`Deque` is what defines those methods, so it would be a bit more clear, but the comment itself is optional, so feel free to leave it as is.

The idea behind my comment: type of variable can provide additional information about how you intend to use it. If the type is `LinkedList`, it tells me only that it is a general container with all the methods that it provides. If type is `Deque` it tells me that you plan to use it as queue or stack and makes it easier to spot situations where it might be misused (say: intended to use it as stack but then looking at random elements inside it via `get(index)`).

Reply:
Right, that makes sense, thank you. I have now changed my DFS implementation to use recursion though so I've removed most of the code in `printAncestors()`.